### ISSUE TITLE
Can't start zowe server successfully with keyring function plus external CA in 1.15.0

### ISSUE THREAD

**Issue by rjmaomao** [zowe/zowe-install-packaging#1664]
I'm trying to explore the new keyring function in Zowe 1.15.0. I have a certificates(label ZoweCert.Z2) which also signed by IBM. I used this with keystone method as external certificates and it worked well. I think this matches the following scenario described in ZWEKRING.
Options:                                                        
  _1. Zowe's certificate is already loaded in RACF database      
     ACTION: Modify the CONNECT(ID(&ZOWEUSER.) LABEL('&LABEL.') 
             command below to match the owner of the desired    
             certificate                                        
             Comment the "Option 2" block below_

Commented out step 'Create Zowe's local CA authority' and 'Option 2' in ZWEKRING JCL.  Ran the job only to create keyring(ZoweKeyring.Z2) and added this certificate(label ZoweCert.Z2) and related IBM root and IBM Intermediate root to the keyring. Also created jwtsecret certificate and connected to keyring.
 
My keyring looks like following now.
 
 _>>>   RACDCERT LISTRING(ZoweKeyring.Z2) ID(IZUSVR)_                                                                                                                    
 ```
Digital ring information for user IZUSVR:                                                                                                                         
   Ring:                                                                          
        >ZoweKeyring.Z2<                                                          
   Certificate Label Name             Cert Owner     USAGE      DEFAULT           
   --------------------------------   ------------   --------   -------           
   ZoweCert.Z2                        ID(IZUSVR)     PERSONAL     YES                                                                                               
   IBM INTERNAL INTERMEDIATE          CERTAUTH       CERTAUTH     NO                                                                                                
   IBM INTERNAL ROOT                  CERTAUTH       CERTAUTH     NO                                                                                                
   zOSMFCA                            CERTAUTH       CERTAUTH     NO                                                                                                
   jwtsecret                          ID(IZUSVR)     PERSONAL     NO 
```
 
Ran zowe-setup-certificates.sh , configured following in zowe-setup-certificates.env file
```
KEYSTORE_DIRECTORY=/zowe/keystore/v1.15.keyring.ex
KEYSTORE_ALIAS=ZoweCert.Z2
ZOWE_KEYRING=ZoweKeyring.Z2
```

Script completed with following output.
_142:/Z2/var/zowert/1.15.0 $ bin/zowe-setup-certificates.sh -p /zowe/keystore/zowe-setup-certificates.env
Log file created: /u/renjing/zowe/logs/zowe-setup-certificates-2020-09-18-02-27-36.log
Loading default variables from /Z2/var/zowert/1.15.0/bin/zowe-setup-certificates.env file.
Loading /zowe/keystore/zowe-setup-certificates.env file and overriding default variables.
error : dig found hostname z2pub.pok.stglabs.ibm.com and IP but IP did not match 9.114.33.20
  ZOWE_IP_ADDRESS    variable value=9.114.33.20
  ZOWE_EXPLORER_HOST variable value=z2pub.pok.stglabs.ibm.com
Creating certificates and keystores... STARTED
Generating certificates for the keyring is skipped.
z/OSMF certificate fingerprint:
Owner: CN=ZPETPLX2ZOSMF.POK.STGLABS.IBM.COM, OU=ZMZ2, O=IBM
     MD5:  47:D1:69:31:0B:8D:2F:20:F5:94:7F:9E:04:3B:37:B5
     SHA1: A5:84:98:67:78:74:DE:00:FD:D6:28:63:1C:07:8A:D0:45:0E:E6:A7
     SHA256: D2:E0:DD:AC:2B:3C:D5:AC:7B:56:84:B7:5C:1E:A3:85:A6:DB:65:36:B6:43:71:02:A5:06:41:C9:82:95:DD:5A
Owner: CN="z/OSMF CertAuth for Security           Domain", OU=ZMZ2
     MD5:  03:9E:E8:36:7D:2C:E0:77:26:2F:81:45:BF:B7:3C:A0
     SHA1: 95:DD:76:17:00:0F:6C:EB:93:D7:96:AD:BA:69:84:41:25:14:FD:C8
     SHA256: 0B:3D:70:38:CC:E3:F3:44:DA:95:88:C5:2C:FA:AA:EA:9A:6E:FD:87:3B:FE:ED:8B:84:D4:33:20:A7:94:E9:10
Creating certificates and keystores... DONE
Detecting external root CA... STARTED
**The ZoweCert.Z2 certificate is self-signed or does not contain its CA chain. If the certificate is externally signed and its root CA is connected to the same keyring then you can manually set the EXTERNAL_ROOT_CA env variable with the ro.**
Detecting external root CA... DONE
Trying to change an owner of the /zowe/keystore/v1.15.keyring.ex.
Owner of the /zowe/keystore/v1.15.keyring.ex changed successfully to the IZUSVR owner._

It mentioned I should **_manually set the EXTERNAL_ROOT_CA env variable with the ro._** Guess there's a typo here, should be **root CA** rather than **ro.** ?

Checking the keystore dir, it was generated, but the local_ca dir is empty, is this correct?
```
135:/zowe/keystore/v1.15.keyring.ex $ ls
ZoweCert.Z2            local_ca               zowe-certificates.env
133:/zowe/keystore/v1.15.keyring.ex $ ls ZoweCert.Z2
ZoweCert.Z2.keystore.jwtsecret.pem  ZoweCert.Z2.truststore.p12
```

The file zowe-certificates.env also looked not right, shouldn't the variable {KEYRING_OWNER} and {KEYRING_NAME} been replaced with real value?
_136:/zowe/keystore/v1.15.keyring.ex $ cat zowe-certificates.env
    KEY_ALIAS=ZoweCert.Z2
    KEYSTORE_PASSWORD="password"
    KEYRING_OWNER="IZUSVR"
    KEYRING_NAME="ZoweKeyring.Z2"
    KEYSTORE="safkeyring:////**${KEYRING_OWNER}/${KEYRING_NAME}**"
    KEYSTORE_TYPE="JCERACFKS"
    TRUSTSTORE="safkeyring:////**${KEYRING_OWNER}**/**${KEYRING_NAME}**"
    EXTERNAL_ROOT_CA=
    ZOWE_APIM_VERIFY_CERTIFICATES=true
    SETUP_APIML_SSO=
    SSO_FALLBACK_TO_NATIVE_AUTH=true
    PKCS11_TOKEN_NAME=
    PKCS11_TOKEN_LABEL=_

Then I continued to create an instance, updated the instance.env file. But when I launch zowe server, API catalog, Discovery Server and Authentication service all failed to start.  Open the flux desktop and tried to logon , got **"Authentication failed for 3 types. Types: ["saf","apiml","zss"]"**. Checkeck from developer tool from browser. Following Responses were given during logon from desktop. 
 
_{"categories":{"saf":{"success":false,"plugins":{"org.zowe.zlux.auth.safsso":{"success":false,"reason":"Unknown","error":{"message":"APIML 503 "},"apiml":true,"zss":true,"sso":false,"canChangePassword":true,"capabilities":{"canGetStatus":true,"canGetCategories":true,"canRefresh":true,"canAuthenticate":true,"canAuthorize":true,"canLogout":true,"canResetPassword":true,"proxyAuthorizations":true,"processesProxyHeaders":true}}}},"apiml":{"success":false,"plugins":{"org.zowe.zlux.auth.safsso":{"success":false,"reason":"Unknown","error":{"message":"APIML 503 "},"apiml":true,"zss":true,"sso":false,"canChangePassword":true,"capabilities":{"canGetStatus":true,"canGetCategories":true,"canRefresh":true,"canAuthenticate":true,"canAuthorize":true,"canLogout":true,"canResetPassword":true,"proxyAuthorizations":true,"processesProxyHeaders":true}}}},"zss":{"success":false,"plugins":{"org.zowe.zlux.auth.safsso":{"success":false,"reason":"Unknown","error":{"message":"APIML 503 "},"apiml":true,"zss":true,"sso":false,"canChangePassword":true,"capabilities":{"canGetStatus":true,"canGetCategories":true,"canRefresh":true,"canAuthenticate":true,"canAuthorize":true,"canLogout":true,"canResetPassword":true,"proxyAuthorizations":true,"processesProxyHeaders":true}}}},"fallback":{"success":true,"plugins":{"org.zowe.zlux.auth.trivial":{"success":true,"capabilities":{"canGetStatus":true,"canRefresh":true,"canAuthenticate":true,"canAuthorize":true,"proxyAuthorizations":false,"canResetPassword":false}}}}},"expms":3600000,"success":false,"defaultCategory":"saf"}_

Also checked Zowe started task job log. 
_2020-09-18 06:45:54.817 <ZWEADS1:TaskBatchingWorker-target_z2pub.pok.stglabs.ibm.com-0:33816938> IZUSVR ERROR (c.n.e.c.ReplicationTaskProcessor) Batch update failure with HTTP status code 403; discarding 1 replication tasks  
2020-09-18 06:45:54.819 <ZWEADS1:TaskBatchingWorker-target_z2pub.pok.stglabs.ibm.com-0:33816938> IZUSVR WARN  (c.n.e.u.b.TaskExecutors) Discarding 1 tasks of TaskBatchingWorker-target_z2pub.pok.stglabs.ibm.com-0 due to permanent error
2020-09-18 07:36:10.861 <ZWEAGW1:https-jsse-nio-0.0.0.0-14522-exec-9:262385> IZUSVR ERROR (o.z.a.g.s.s.z.ZosmfServiceFacade) ZWEAG108E z/OSMF instance 'zosmf'  not found or incorrectly configured.      
2020-09-18 07:36:11.205 <ZWED:17040342> IZUSVR WARN (_zsf.auth,webauth.js:328) ZWED0003W - CUnInn2jZk5mYNNborKOc1tFTnqTNacy: Session security call authenticate failed for auth handler org.zowe.zlux.auth.safsso. Plugin response: {"success"
:false,"reason":"Unknown","error":{"message":"APIML 503 "},"apiml":true,"zss":true,"sso":false,"canChangePassword":true}
2020-09-18 07:36:11.205 <ZWED:17040342> IZUSVR INFO (_zsf.auth,webauth.js:325) ZWED0070I - CUnInn2jZk5mYNNborKOc1tFTnqTNacy: Session security call authenticate succesful for auth handler org.zowe.zlux.auth.trivial. Plugin response: {"succ
ess":true}
2020-09-18 07:36:31.020 <ZWED:17039492> IZUSVR WARN (_zsf.auth,webauth.js:328) ZWED0003W - CUnInn2jZk5mYNNborKOc1tFTnqTNacy: Session security call authenticate failed for auth handler org.zowe.zlux.auth.safsso. Plugin response: {"success"
:false,"reason":"Unknown","error":{"message":"APIML 503 "},"apiml":true,"zss":true,"sso":false,"canChangePassword":true}
2020-09-18 07:36:31.022 <ZWED:17039492> IZUSVR INFO (_zsf.auth,webauth.js:325) ZWED0070I - CUnInn2jZk5mYNNborKOc1tFTnqTNacy: Session security call authenticate succesful for auth handler org.zowe.zlux.auth.trivial. Plugin response: {"succ
ess":true}_
 
I have several questions.

1.  Is my way of using keyring plus external CA correct?
2.  Is it normal the local_ca dir under keystore directory empty if I use keyring method? I checked log file and saw information about creating certificates in local_ca dir
_Calling keytool -export -v -alias localca -file /zowe/keystore/v1.15.keyring.ex/local_ca/localca.cer -keystore /zowe/keystore/v1.15.keyring.ex/local_ca/localca.keystore.p12 -rfc -keypass local_ca_password -storepass local_ca_password -storetype PKCS12
Certificate stored in file </zowe/keystore/v1.15.keyring.ex/local_ca/localca.cer>
keytool returned: 0_
 
3.  My certificate ZoweCert.Z2 have a complete certificates chain including the IBM root and IBM Intermedia, do I still need to add EXTERNAL_ROOT_CA manually? I now see this value can be added in zowe-certificates.env,  Is IBM INTERNAL ROOT the correct value to use? Do I need to re-create instance after updating this value?
4. Is {KEYRING_OWNER} and {KEYRING_NAME} correct in zowe-certificates.env? Should it be updated to real value?
5. How to fix the APIML 503 error during desktop logon?

Thanks!

**Comment by stevenhorsman** [comment]
Copying @vit-tomica's response for public references:

1.  Is my way of using keyring plus external CA correct?
     **The keyring setup with external certificates looks OK.**

2.  Is it normal the local_ca dir under keystore directory empty if I use keyring method? I checked log file and saw information about creating certificates in local_ca dir
    **It is normal that local_ca dir is empty. Internally, the config script creates a temporary certificate (named as local_ca) that only helps to set up SSO for Zowe Desktop using z/osmf JWT. This is the only purpose of the temporary certificate which is then deleted at the end of the config process. This is why you see it in the log.**
 
3.  My certificate ZoweCert.Z2 have a complete certificates chain including the IBM root and IBM Intermedia, do I still need to add EXTERNAL_ROOT_CA manually? I now see this value can be added in zowe-certificates.env,  Is IBM INTERNAL ROOT the correct value to use? Do I need to re-create instance after updating this value?
     **Yes, you should manually set the EXTERNAL_ROOT_CA variable with the "IBM INTERNAL ROOT" value. Since the value contains white spaces it should be enclosed in quotes. Anyway, I'm not sure whether a label containing white spaces will work correctly when set to EXTERNAL_ROOT_CA variable.
You don't have to re-create the instance dir.** 
 
4. Is {KEYRING_OWNER} and {KEYRING_NAME} correct in zowe-certificates.env?
   **This is correct. ${KEYRING_OWNER} is substituted by the KEYRING_OWNER env variable in the same file, similarly the {KEYRING_NAME} var.**

5. How to fix the APIML 503 error during desktop logon? 
    **I'd suggest turning on debug mode in APIML to see more messages. Set this variable to debug in the <rootdir>/components/api-mediation/bin/start.sh**

**I can see an error message in appServer log:
2020-09-18 06:45:23.288 <ZWED:17039492> IZUSVR INFO (_zsf.bootstrap,webserver.js:191) ZWED0072I - Using Certificate: safkeyring:////IZUSVR/ZoweKeyring.Z2&ZoweCert.Z2
2020-09-18 06:45:23.290 <ZWED:17039492> IZUSVR WARN (_zsf.bootstrap,webserver.js:156) ZWED0148E - Exception thrown when reading SAF keyring, e= TypeError: R_datalib call failed: function code: 01, SAF rc: 8, RACF rc: 8, RACF rsn: 44

The message indicates that the ZoweCert.Z2 cert is not found in the keyring. I'm not sure why this happens because the keyring looks correct.**

**Comment by rjmaomao** [comment]
Thank you @stevenhorsman and @vit-tomica 

You mentioned it's normal that local_ca dir is empty. I want to confirm it should be empty only for keyring configuration, right? Regardless of whether you are using self-signed certificates and external certificates keyring, local_ca dir is always empty. But for unix file configuration, end user should have following files generated in local_ca dir, is that correct?
```
138:/zowe/keystore/v1.14.ex.sso/local_ca $ ls
extca.1.cer     extca.1.cer-ebcdic    extca.2.cer     extca.2.cer-ebcdic    localca.cer    localca.cer-ebcdic    localca.keystore.p12

```
I did as you suggested.

I added EXTERNAL_ROOT_CA to zowe-certificates.env and recycled the server, but still got the same error. 
_/zowe/keystore/v1.15.keyring.ex/ zowe-certificates.env
  EXTERNAL_ROOT_CA="IBM INTERNAL ROOT"_

I enabled debug to APIML by editing 

_/Z2/var/zowert/1.15.0/components/api-mediation/bin/ start.sh
To activate `debug` mode, set LOG_LEVEL=debug (in lowercase)
LOG_LEVEL=**debug**_

I noticed the debug info was not in appServer.log, but they were in started task joblog. Is this by design? 

Still see ZWED0148E message.
_2020-09-22 03:53:09.741 <ZWED:67371713> IZUSVR INFO (_zsf.bootstrap,webserver.js:191) ZWED0072I - Using Certificate: safkeyring:////IZUSVR/ZoweKeyring.Z2&ZoweCert.Z2
 2020-09-22 03:53:09.743 <ZWED:67371713> IZUSVR WARN (_zsf.bootstrap,webserver.js:156) ZWED0148E - Exception thrown when reading SAF keyring, e= TypeError: R_datalib call failed: function code: 01, SAF rc: 8, RACF rc: 8, RACF rsn: 44_


The error code SAF rc 8, RACF8 rc, RACF rsn 44 indicates No trusted certificate found. But I did a racdcert list command, it's showed in TRUST status.                                                                       
_Digital certificate information for user IZUSVR:                                                                                                                     
  Label: ZoweCert.Z2                                                               
  Certificate ID: 2QbJ6eTi5dnplqaFw4WZo0vp8kBA                                     
  Status: TRUST                                                                    
  Start Date: 2020/02/10 01:00:00                                                  
  End Date:   2022/05/14 23:59:59                                                  
  Serial Number:                                                                   
       >012AA5<                                                                    
  Issuer's Name:                                                                   
       >CN=IBM INTERNAL INTERMEDIATE CA.O=International Business Machines Cor<     
       >poration.C=US<                                                             
  Subject's Name:                                                                  
       >renjbj@cn.ibm.com.CN=z2pub.pok.stglabs.ibm.com.OU=IBM Pough<     
       >keepsie.O=ibm.com.L=New York, NY.SP=New York, NY.C=US<                     
  Subject's AltNames:                                                              
    Domain: z2pub.pok.stglabs.ibm.com                                              
  Signing Algorithm: sha256RSA                                                     
  Key Usage: HANDSHAKE                                                             
  Key Type: RSA                                                                    
  Key Size: 2048                                                                   
  Private Key: YES                                                                 
  Ring Associations:                                                               
    Ring Owner: IZUSVR                                                             
    Ring:                                                                          
       >ZoweKeyring.Z2<_                                                            
                                                                                   
***                                                                                

Also saw error message related to certificates _localca_ and _IBM INTERNAL ROOT_. The _localca_ certificates wasn't created because I'm using this external CA scenario that didn't require creating localca. But for "IBM INTERNAL ROOT", it's specified in EXTERNAL_ROOT_CA, and it existed in the system, not sure why can't be found.

 _2020-09-22 03:53:10.420 <ZWED:17040135> IZUSVR WARN (org.zowe.terminal.proxy:tn3270data,terminalProxy.js:681) ZWED0144W - Error when reading **file=safkeyring:////IZUSVR/ZoweKeyring.Z2&localca. Error=ENOENT: no such file or directory**, open 'safkeyring:////IZUSVR/ZoweKeyring.Z2&localca'
 2020-09-22 03:53:10.420 <ZWED:17040135> IZUSVR WARN (org.zowe.terminal.proxy:tn3270data,terminalProxy.js:681) ZWED0144W - Error when reading file=**safkeyring:////IZUSVR/ZoweKeyring.Z2&IBM INTERNAL ROOT. Error=ENOENT: no such file or directory,** open 'safkeyring:////IZUSVR/ZoweKeyring.Z2&IBM INTERNAL ROOT'_

**Comment by vit-tomica** [comment]
@rjmaomao Everything that you mentioned about the local_ca directory is correct. It should be empty for keyring configuration.

The appServer.log file contains only messages related to Zowe App Framework. Debug mode was activating for APIML so it is fine not having debug messages in the appServer.log.

I think I start to understand better what's happening here:
`_2020-09-22 03:53:09.741 ZWED:67371713 IZUSVR INFO (_zsf.bootstrap,webserver.js:191) ZWED0072I - Using Certificate: safkeyring:////IZUSVR/ZoweKeyring.Z2&ZoweCert.Z2`
The message indicates that the certificate is OK and is being used. The proof is that you can actually get to the Zowe Desktop login screen so you should see the ZoweCert.Z2 certificate in the browser. The fact that you can't login to Zowe Desktop is a different problem related to APIML.

The other message ` 2020-09-22 03:53:09.743 ZWED:67371713 IZUSVR WARN (zsf.bootstrap,webserver.js:156) ZWED0148E - Exception thrown when reading SAF keyring, e= TypeError: R_datalib call failed: function code: 01, SAF rc: 8, RACF rc: 8, RACF rsn: 44` is related to the localca certificate that actually does not exist in the keyring because you commented that out in the JCL. The Zowe desktop hardcodes the localca certificate in its config scripts and expects that the localca always exists. Historically, when using certificates in USS keystores, the localca certificates is always created (even when external certificates are used) because it acts as a Zowe local CA which may be useful, e.g. a user can easily generate a certificate using the local CA for a third party service that wants to register to APIML. 
So since the localca cert is always created for keystores, we do the same thing for the keyrings in order not to break anything.
Long story short, if you let the ZWEKRING JCL to generate a localca cert and connect it to the keyring then the message should disappear. I believe there should be no instructions in the doc or JCL that instructs to comment out the localca cert creation. In your case, if the localca cert does not exist then it should not affect functionality of Zowe Desktop but the error message is written and it may be confusing. I'm thinking what we could do about that...

I believe the actual problem is in the APIML but I'm not sure where yet. I'm going to pull someone in from APIML team to look at that.

**Comment by rjmaomao** [comment]
@vit-tomica 
Hi Vit,

If I understand correctly , certificate 'safkeyring:////IZUSVR/ZoweKeyring.Z2&localca'   shouldn't be required when using external CA. But since it's hardcoded we are seeing this error. It may not impact the functionality. I saw you have another issue opened. 

For certificate 'safkeyring:////IZUSVR/ZoweKeyring.Z2&IBM INTERNAL ROOT', it may be a bug within APIML. And this may be the cause that I can't logon to the desktop. You will have APIML team to look into this issue. 


I also tried local CA scenario with ZWEKRING JCL. I didn't comment out anything, only updated HOSTNAME, IPADDRES, ZOWERING, LABEL, ROOTZFCA, LOCALCA, as well as field for SUBJECTSDN to fit our environment.  I got RC 12 submitting ZWEKRING, and saw following error in job log. 

```
IEC020I 001-4,ZWEKRING,RUN ,SYSTSIN ,JES
IEC020I EROPT IS 'ABE' OR NOT SPECIFIED

IKJ56600I UNRECOVERABLE COMMAND SYSTEM ERROR
```

I opened following issue. Could you please also help to take a look? 

https://github.com/zowe/zowe-install-packaging/issues/1666


The default value of LOCALCA in ZWEKRING is 'localca'. I want to confirm that I can update this value, right? The code should pick up the new value rather than using hardcoded localca as we see when using external CA, is that correct?

**Comment by vit-tomica** [comment]
@rjmaomao
I'll take a look at the JCL issue.

Please use the default 'localca' value. If you specify a different value then you'll very likely get the ` R_datalib call failed: function code: 01, SAF rc: 8, RACF rc: 8, RACF rsn: 44` error again. I'll open a new issue for this.

**Comment by vit-tomica** [comment]
I opened a new issue regarding the 'localca' value - https://github.com/zowe/zowe-install-packaging/issues/1669

**Comment by vit-tomica** [comment]
@rjmaomao Based on the comment [here](https://github.com/zowe/zowe-install-packaging/issues/1666#issuecomment-698788944), you managed to successfully start Zowe server. Do you still have the issue with the APIML that prevents you to logon to Zowe Desktop? Or did it disappear?

**Comment by rjmaomao** [comment]
@vit-tomica Hi Vit, I was able to deploy zowe with local ca after deleting the comments you mentioned, but still failed to use external ca. 
 ```
2020-09-22 03:53:10.420 ZWED:17040135 IZUSVR WARN (org.zowe.terminal.proxy:tn3270data,terminalProxy.js:681) ZWED0144W - Error when reading file=safkeyring:////IZUSVR/ZoweKeyring.Z2&IBM INTERNAL ROOT. Error=ENOENT: no such file or directory, open 'safkeyring:////IZUSVR/ZoweKeyring.Z2&IBM INTERNAL ROOT'
```

**Comment by vit-tomica** [comment]
@rjmaomao Hi Jing, the warning message does not indicate any real problem according to the discussion [here](https://github.com/zowe/zlux/issues/512). Is there any other message in logs that indicates a failure?

**Comment by rjmaomao** [comment]
Hi @vit-tomica , I'm a little confused here. There are two errors in log, one complain about **safkeyring:////IZUSVR/ZoweKeyring.Z2&localca**, you opened [defect](https://github.com/zowe/zlux/issues/512) for this one.  And according to the discussion in the defect this won't impact functionality. If I didn't commented out generation of localca in JCL this error should be gone. Even if I comment out localca generation code, this error is just a warning in the log.

But the second one is asking for  **ZoweKeyring.Z2&IBM INTERNAL ROOT**.  This one is for external ca, unlike localca, which doesn't require this internal root, when using external ca, we need to specify IBM root and IBM Intermediate Root. You mentioned in your previous comments 
```
These two messages are probably caused by a bug. I think there is code that expects 
that safkeyring:////IZUSVR/ZoweKeyring.Z2&IBM INTERNAL ROOT is a regular file which 
is not true and so the message is issued. I don't know whether there is any other impact of the bug.

I believe the actual problem is in the APIML but I'm not sure where yet. I'm going to pull someone 
in from APIML team to look at that."
```
I haven't try on 1.16, but I think I will still get error about ZoweKeyring.Z2&IBM INTERNAL ROOT when using external ca configuration.

```
 2020-09-22 03:53:10.420 ZWED:17040135 IZUSVR WARN (org.zowe.terminal.proxy:tn3270data,terminalProxy.js:681) ZWED0144W - Error when reading file=safkeyring:////IZUSVR/ZoweKeyring.Z2&localca. Error=ENOENT: no such file or directory, open 'safkeyring:////IZUSVR/ZoweKeyring.Z2&localca'
 2020-09-22 03:53:10.420 ZWED:17040135 IZUSVR WARN (org.zowe.terminal.proxy:tn3270data,terminalProxy.js:681) ZWED0144W - Error when reading file=safkeyring:////IZUSVR/ZoweKeyring.Z2&IBM INTERNAL ROOT. Error=ENOENT: no such file or directory, open 'safkeyring:////IZUSVR/ZoweKeyring.Z2&IBM INTERNAL ROOT'
```

Please let me know if I have any misunderstanding. Thanks you for help on this!

**Comment by vit-tomica** [comment]
@rjmaomao Hi Jing, my understanding (based on the [issue ](https://github.com/zowe/zlux/issues/512#issuecomment-702389997)) is that we should stop looking at that message in general (no matter if we see 'localca' or 'IBM INTERNAL ROOT' in the message) because it does not impact keyring functionality. The message is confusing and misleading. 

I would suggest to get back a little and to recap what works or does not work from user's perspective.
Can you see please confirm whether you can see the following?
 - Gateway landing page (Can you click on the API catalog link and try to logon to the API catalog, please? )
![image](https://user-images.githubusercontent.com/55885025/96701217-a20e5a00-1390-11eb-923e-b55f9dc7fa2a.png)

- Zowe Desktop login page 
![image](https://user-images.githubusercontent.com/55885025/96702059-a7b86f80-1391-11eb-99ec-9dff70a79e80.png)


Thanks for your patience!

**Comment by vit-tomica** [comment]
Hi @rjmaomao, I worked on an issue recently that had very similar symptoms to this issue. The problem was that Zowe certificate was missing the `Client Authentication (OID 1.3.6.1.5.5.7.3.2)` attribute which is one of the Extended Key Usage attributes that can be added to a generated certificate. I think it could be worth checking whether your certificate has that attribute.

**Comment by MarkAckert** [comment]
Due to inactivity, I will be closing this issue. If you think it should be re-opened, please do so and if possible with an updated description of the problem. Thank you!
